src/ps.rs

```diff
@@ -250,444 +132,450 @@ fn do_create_snapshot(
-        hostname: &hostname,
-        timestamp: &system.get_timestamp(),
-        version: &system.get_version(),
-        flat_data: !opts.new_json,
-        epoch: system.get_boot_time() - EPOCH_TIME_BASE,
-        opts,
-    };
-
-    match collect_data(system, &print_params) {
-        output::Value::A(elts) => {
-            for i in 0..elts.len() {
-                output::write_csv(writer, elts.at(i));
-            }
-        }
-        obj @ output::Value::O(_) => {
-            output::write_json(writer, &obj);
-        }
-        output::Value::E() => {
-            // interrupted, don't print anything
-        }
-        _ => {
-            panic!("Should not happen")
-        }
-    }
-}
-
-// If this returns an output::Value::O then that is an object to write (eg JSON), otherwise it must
-// be an output::Value::A and each should be written individually (eg CSV), or it is
-// output::Value::E, in which case we were interrupted.  The first two cases are controlled by
-// print_params.flat_data.
-
-fn collect_data(
-    system: &dyn systemapi::SystemAPI,
-    print_params: &PrintParameters,
-) -> output::Value {
-    match do_collect_data(system, print_params) {
-        Ok(output::Value::A(mut elts)) => {
-            if elts.len() == 0 && print_params.opts.always_print_something {
-                elts.push_o(make_heartbeat(&print_params))
+    match collect_sample_data(system, opts) {
+        Ok(Some(sample_data)) => {
+            if opts.new_json {
+                let o = output::Value::O(format_newfmt(&sample_data, system, opts));
+                output::write_json(writer, &o);
+            } else {
+                let mut elements = format_oldfmt(&sample_data, system, opts).take();
+                if elements.len() == 0 && opts.always_print_something {
+                    elements.push(output::Value::O(make_oldfmt_heartbeat(system)))
+                }
+                for e in &elements {
+                    output::write_csv(writer, e);
+                }
             }
-            output::Value::A(elts)
         }
-        Ok(obj @ output::Value::O(_)) => obj,
-        Ok(empty @ output::Value::E()) => empty,
-        Ok(_) => {
-            panic!("Should not happen")
+        Ok(None) => {
+            // Interrupted, do not print anything
         }
         Err(error) => {
-            let mut hb = make_heartbeat(&print_params);
-            hb.push_s("error", error);
-            if print_params.flat_data {
-                output::Value::A(output::Array::from_vec(vec![output::Value::O(hb)]))
+            if opts.new_json {
+                let mut envelope = output::newfmt_envelope(system);
+                envelope.push_a("errors", output::newfmt_one_error(system, error));
+                output::write_json(writer, &output::Value::O(envelope));
             } else {
-                output::Value::O(hb)
+                let mut hb = make_oldfmt_heartbeat(system);
+                hb.push_s("error", error);
```

Comment:
Hm, this does not need a +oldnames annotation, which suggests that test coverage is not as good as it needs to be (we're not checking ps.rs).